### DISCUSSION TITLE
Opinions about the game, the development and what it currently looks like.

### DISCUSSION THREAD

**Discussion by RisingLeaf** [endless-sky/endless-sky#10412] [General]
First of all big disclaimer: These are **my** opinions, they do not need to be yours and you are free to disagree. I am not trying to dictate my opinions here I am trying to give you whoever you are an impulse to think.

Second disclaimer: Not all of these opinions are criticism, some just state what's already happening and what I like, so I would welcome it if you refrain from telling me that something is already in place.

Third disclaimer: Nothing here is directed at specific people but only at the process as a whole.

Now with this in mind, feel free to comment.

## Opinions regarding ES as a game:

1. ES is an open world sandbox sci-fi space game, the player should be able to choose a path and playstyle that fits them, this means ES should support all of its sub gameplays, these are combat, mining, trading, missions and the story equally and each of them should provide equal opportunities for growth of the player. Equally meaning not entirely equal but at least a bit balanced against each other.

2. Plugins are a central part of ES and should be treated as such. Not every feature needs to be designed for the vanilla game and other features should be welcome too. Doing so opens a much broader sky of possibilities for ES and a lot more room for tryouts and failures, which will in return benefit the main game when ideas show to work excellently.

## Regarding ES as a game engine:

1. One of the games major strengths is the ability to run on nearly any platform, this sentiment should be held in high regard so new features should try to be efficient and not be on the cutting edge of progress. A lot of ES players rely on this fact and this therefore helps to keep the player base from which our contributors come too.

2. ES is made to be accessible, this means code and data formats should be written so that it is easy to understand and get into for beginners. This helps keeping the project alive as senior contributors will not stay forever.

## Regarding the process of development:

1. Every contributor has good intentions and wants to improve the game, this is independent of their status inside the team. But every person contributing to the game has a limited amount of time, therefore game development happens inside a limited time pool. The time spent developing should therefore be a non trivial concern when talking about changing contributions. A thank you to the contributor after merging a PR would by the way help everyone to feel valued and it's not really exhausting to do so. Of course a thank you to maintainers is also always in place.

2. We have structures and rules here, so they need to be written down. If we did not have them we wouldn't need to write them down, but we have them so it is necessary. This should also include a list of people with statuses above standard contributor.

3. The devs, as we call them, should not be called devs, the name implies they are the ones developing the game but this is wrong, while a dev might open a PR their role is to decide what goes in. They should therefore call themselves maintainers.

4. Not all maintainers should look at all areas of the game. The process of developing is a lot easier when maintainers keep to their area of expertise.

5. The game doesn't need to be perfect, fixes can be deployed at any time during unstable releases so there should be no reason to hesitate too much with merge. Doing so allows for faster progress as well a more enjoyable development experience and the ability to trial and error will only improve the core game. Of course when we get closer to a release it makes sense to be more careful as we want to avoid deploying patches.

6. There is absolutely no need to have hidden channels on the discord server, it should be common knowledge what the main contributors talk about. Acceptable restrictions might be write permissions or the moderation room.

## Conclusion

These are my personal views, if you read them you may decide for yourself which ones of these are already in existence and which are not and maybe you find yourself agreeing with one point and seeing it is not done yet and then my goal is reached.

Thank you for reading 🙏 (if you just read the last line you are not included 😤)

**Comment by Hecter94** [comment]
Just a general overview of which parts I'm in agreement about, which I disagree on, and which I'm conflicted about, or think there are valid arguments both for and against.

<details>
<summary>Agree</summary>

1. ES is an open world sandbox sci-fi space game, the player should be able to choose a path and playstyle that fits them, this means ES should support all of its sub gameplays, these are combat, mining, trading, missions and the story equally and each of them should provide equal opportunities for growth of the player. Equally meaning not entirely equal but at least a bit balanced against each other.
2. ES is made to be accessible, this means code and data formats should be written so that it is easy to understand and get into for beginners. This helps keeping the project alive as senior contributors will not stay forever.
3. We have structures and rules here, so they need to be written down. If we did not have them we wouldn't need to write them down, but we have them so it is necessary. This should also include a list of people with statuses above standard contributor.
</details>

<details>
<summary>Disagree</summary>

1. The devs, as we call them, should not be called devs, the name implies they are the ones developing the game but this is wrong, while a dev might open a PR their role is to decide what goes in. They should therefore call themselves maintainers.
2. Not all maintainers should look at all areas of the game. The process of developing is a lot easier when maintainers keep to their area of expertise.
3. The game doesn't need to be perfect, fixes can be deployed at any time during unstable releases so there should be no reason to hesitate too much with merge.
4. There is absolutely no need to have hidden channels on the discord server, it should be common knowledge what the main contributors talk about. Acceptable restrictions might be write permissions or the moderation room.
</details>

<details>
<summary>Conflicted</summary>

1. Plugins are a central part of ES and should be treated as such. Not every feature needs to be designed for the vanilla game and other features should be welcome too. Doing so opens a much broader sky of possibilities for ES and a lot more room for tryouts and failures, which will in return benefit the main game when ideas show to work excellently.
2. One of the games major strengths is the ability to run on nearly any platform, this sentiment should be held in high regard so new features should try to be efficient and not be on the cutting edge of progress. A lot of ES players rely on this fact and this therefore helps to keep the player base from which our contributors come too.
3. Every contributor has good intentions and wants to improve the game, this is independent of their status inside the team. But every person contributing to the game has a limited amount of time, therefore game development happens inside a limited time pool. The time spent developing should therefore be a non trivial concern when talking about changing contributions.
</details>

I want to avoid the conversation getting bogged down in a disagreement of ideals, so I've simply stated which sentiments I support and which I do not.
However, if you would like to know my reasoning for supporting or opposing something or my reason for being conflicted on a point, feel free to ask, and I will be happy to answer.

As a starting point, on the things we agree on, I would be interested in knowing:

1. In your opinion, what are the different "playstyles" that Endless Sky currently supports, and do you feel that some are not well balanced regarding each other? If so, which, and do you have any suggestions for resolving it?
2. In a similar vein, are there any parts of Endless Sky's code or data file structure that you feel are too difficult to understand for beginners? If so, do you have any suggestions for how to fix that?
3. Finally, are there rules or structures that you feel need to be more visible, and do you have any suggestions for how they could be made more visible?

**Reply by RisingLeaf** [reply]
> 1. In your opinion, what are the different "playstyles" that Endless Sky currently supports, and do you feel that some are not well balanced regarding each other? If so, which, and do you have any suggestions for resolving it?

Main playstyles I think of:
- solo or small fleet, plays a lot of story, enjoys a challenge in combat
- large scale trading, big fleet, likes to make lots of money
- mining, this one is currently the least profitable playstyle imo, but I see it getting better (gegno, tractor beams, etc.)
- flying jobs or just doing jobs in general, variable fleet size
- power accumulation, looking for best builds, capturing 
- explorer, mostly solo ship, likes to "see stuff" not so much story as experience the galaxy (this is me, not sure how many others are playing like this) more profitable than mining, but I think this one has to be mixed with other stuff to be profitable 

> 2. In a similar vein, are there any parts of Endless Sky's code or data file structure that you feel are too difficult to understand for beginners? If so, do you have any suggestions for how to fix that?

I don't actually see a problem here, I think the syntax is pretty good, wiki needs some filling up though.

> 3. Finally, are there rules or structures that you feel need to be more visible, and do you have any suggestions for how they could be made more visible?

There seems to be a lot of rules or guidelines around that are not directly visible to the average contributor, such are:
Who is going to merge my PR and what person's opinion is important for a merge.
How do people get selected into more permissive roles.
What are these people allowed to do.
Is it mandatory to enter the discord server (no, but for certain stuff almost yes)
Who are the core members? (remember the campfire chat questions)
What is the standard release cycle?
What are rfcs? (Most don't know they exist)
Etc. etc.

I would also be interested in the conflicted section of your comment, I expect for the disagreement section it's mostly a collision of ideals?

**Reply by Hecter94** [reply]
> I don't actually see a problem here, I think the syntax is pretty good, wiki needs some filling up though.

Do you feel that creating the wiki as a separate [repository](https://github.com/endless-sky/endless-sky-wiki) that anyone can PR to is a good step towards resolving this? Now that we have that in play, anyone who feels the Wiki lacks useful information can PR it directly to the wiki. 
This was a relatively recent change, but I hope that over time, it should help to significantly expand the topics that the Wiki covers. 

> There seems to be a lot of rules or guidelines around that are not directly visible to the average contributor, such are:
Who is going to merge my PR and what person's opinion is important for a merge.
How do people get selected into more permissive roles.
What are these people allowed to do.
Is it mandatory to enter the discord server (no, but for certain stuff almost yes)
Who are the core members? (remember the campfire chat questions)
What is the standard release cycle?
What are rfcs? (Most don't know they exist)
Etc. etc.

Yes, I see your point. While some of these are defined, not all are, and even with the ones that are defined, there's no easily visible place to get the answers to all of these questions. 
I agree with you that it would be beneficial to discuss a few of these points, conclude how exactly we want them to be defined, and have that posted somewhere visible. 
For RFCs particularly, to single that point out, I believe we still have to decide if we even want to use RFCs. They're not currently a part of our process in any meaningful way. 
 
> I would also be interested in the conflicted section of your comment, I expect for the disagreement section it's mostly a collision of ideals?

As for the sections I'm conflicted about, going through each in order,

> 1. Plugins are a central part of ES and should be treated as such. Not every feature needs to be designed for the vanilla game and other features should be welcome too. Doing so opens a much broader sky of possibilities for ES and a lot more room for tryouts and failures, which will in return benefit the main game when ideas show to work excellently.

I fully recognize that a large part of our player base shows quite a passion for plugins, either creating or playing them. Thus, while I recognize that expanding the capability of the engine to better support plugin content has clear benefits, I also have a genuine concern regarding the additional maintenance overhead added by features that we don't ever plan to use in Vanilla. 
New features and functionality don't exist in a vacuum; they need to be tested, maintained, and supported in perpetuity, which is a maintenance burden that inevitably hinders further development. 

I'm conflicted about what the stance should be on that, as I recognize there are valid arguments both for and against.
Personally, I feel this is something that simply has to be evaluated on a case-by-case basis. The benefits that any new features could bring to plugin authors must be weighed against whether the extra maintenance burden is worth the additional functionality. 


> 2. One of the games major strengths is the ability to run on nearly any platform, this sentiment should be held in high regard so new features should try to be efficient and not be on the cutting edge of progress. A lot of ES players rely on this fact and this therefore helps to keep the player base from which our contributors come too.

While this is certainly true, we should make a reasonable effort to continue supporting as many platforms as possible. There are limitations to this. Occasionally, we will need to drop support for older platforms or push our system requirements past what older devices may be able to handle if we want to continue expanding upon the experience that Endless Sky can provide. 
It raises the question, "What percentage of our player base are we okay with alienating to improve the experience for the remainder?" which is a challenging question to answer and my main source of conflict in this regard. 

I believe we shouldn't feel obligated to continue supporting every platform in perpetuity, but where exactly we should draw the line isn't a question I can answer. 

> 3. Every contributor has good intentions and wants to improve the game, this is independent of their status inside the team. But every person contributing to the game has a limited amount of time, therefore game development happens inside a limited time pool. The time spent developing should therefore be a non trivial concern when talking about changing contributions.

My main source of conflict here largely comes down to the use of "non-trivial". 
While I do, of course, agree that time spent developing should be considered, I also believe that, ultimately, we have to take whatever is best for the end product. Even if an individual spends a lot of time working on something, if what they worked on doesn't benefit Endless Sky, I don't believe it should be accepted solely because it had a lot of work put into it. 

In a similar vein, if we have a situation where there are, for example, two competing reworks for the same system, one of which has been worked on for over a year whereas the other was done in an afternoon. We should, of course, *consider* the fact that one has had substantially more work put into it, but ultimately, we have to take the one that will be the biggest improvement to the end product, even if that means taking the work that had substantially less time put into it.

So, my main source of conflict here largely comes down to what "non-trivial" means in this context. I agree that it should be considered, but I don't believe it should ever come before the more important question of "What is best for the end product?"

**Reply by RisingLeaf** [reply]
Thanks for the insight into the conflicted section, I see your points and I find all of them reasonable if I read correctly you are mostly worried about the extent of such ideals.

On the other two:
I think having a separate repo for the wiki is very good as we can now make an additional PR to this repo a requirement, so it treats the effect of stuff missing, I don't think it really helps with stuff that's already missing, because the number of people who know about it is very small and they often don't know that it's not documented, it's difficult to fix in that regard.


As for the part on rfcs , I agree whatever way we go with rfcs should be clear and documented, removing them might not be a bad idea, while I like the principle, they are just not really used(two exceptions)

**Comment by tibetiroka** [comment]
I generally agree with what you say here, but there are two things I'd like to mention:

> The devs, as we call them, should not be called devs, the name implies they are the ones developing the game but this is wrong, while a dev might open a PR their role is to decide what goes in. They should therefore call themselves maintainers.

While I wouldn't be against the name, GitHub has a role called Maintainer, but devs have the role one below that iirc. This could make things confusing for people who know GitHub's role system. I can't think of a better name, though.

> Not all maintainers should look at all areas of the game. The process of developing is a lot easier when maintainers keep to their area of expertise.

Ideally, yes. There are absolutely areas of the game I'm not completely comfortable working with. We are however more or less forced to look at them, as we don't have enough people to be able to effectively dedicate portions of the code to them. (Though we do have this separation working for the code/content/balance/art split, but not anything more detailed.)

**Reply by RisingLeaf** [reply]
I was not aware that maintainer is already defined otherwise. I still think the name developer all by itself is misleading though, one alternative could be core developer or some other prefix to differentiate.


For the second part I don't see an actual issue with how it is right now, of course more rigid areas of authority would be better, but as you said you are not enough people to do so.

**Comment by mOctave** [comment]
I pretty much agree with all of these points as well, although I do share some of Hecter's concerns.

> There is absolutely no need to have hidden channels on the discord server, it should be common knowledge what the main contributors talk about. Acceptable restrictions might be write permissions or the moderation room.

I'm conflicted about this point. I think that ideally all dev decisions except for moderation and un-PRed secret lore that the wider playerbase isn't supposed to know about yet should occur in publicly visible channels, or have a public transcript or summary provided. I do appreciate, though, that sometimes a content creator may want to run a vague idea past a reasonably large group of trusted individuals without sharing their ideas with everyone on the discord server, so I think having hidden channels like #high-class-content are helpful, especially as the server (and the number of off-topic discussions) grows.

In short, I think there should be places to discuss things in private, just not places to make decisions in private.

> The game doesn't need to be perfect, fixes can be deployed at any time during unstable releases so there should be no reason to hesitate too much with merge. Doing so allows for faster progress as well a more enjoyable development experience and the ability to trial and error will only improve the core game. Of course when we get closer to a release it makes sense to be more careful as we want to avoid deploying patches.

I think this is actually a really important point. We want ES to be as good a game as possible, of course, but in _any_ system like the ES development cycle, imperfect things are going to get merged—why else do we have bug fixes? The warnings on continuous builds are pretty severe, but for the most part the builds themselves are as stable as a release build. If that's the case, what's the point of having release builds at all?

I think having a couple of months between unstable releases gives us the opportunity to be a bit more relaxed with continuous builds. Despite the warnings, we _should_ try to make things going in the game not crash computers, corrupt saves, or eat kittens, and they should be as good as possible. But "as good as possible" is never going to be the same as "perfect," which I feel is something that we're losing sight of. Every release, quality controls become slightly tighter, and more is required from new additions. This is a good thing, but it behaves like a build-up to a 1.0 release, despite the fact that a lot of things are missing before we can actually push a 1.0 release. Thus, every release, we reduce the number of people with the skill to contribute to the game, while still trying to add more content of all sorts. This seems counter-intuitive.

I strongly agree with Leaf here. We should be more free with the merge button in the first half or so of a development cycle, then tighten things up a bit for the second half and use it to add a bit more and iron out bugs in the first half of the cycle. Then stable releases should (mostly) stop adding features and instead finish ironing out any leftover bugs.

Although I know some people disagree, I think the number of open PRs ES has is actually a big problem, because it concentrates reviews on new PRs and leaves the old ones to rot. Being freer with the merge button would allow for PRs to get merged before their creators or reviewers abandoned them, a state that about a third of our PRs currently appear to be in.

I think the two biggest questions that need regarding this are these:

1. Are the number of potential contributors (especially for art) we lose from raising our expectations fewer than the number of potential contributors we gain from increasing the quality of the game and attracting more players?
2. Is it better to merge a buggy feature early and refine it for the next six months, or to not merge it until six years from now?

**Reply by Azure3141** [reply]
I agree with the sentiment behind relaxed merge conditions, but I worry that it could lead to situations where we merge buggy / unfinished PRs, and then that creates the obligation for someone (possibly the original author, but also likely someone else like a developer given that authors who leave PRs in a buggy/incomplete state likely won't come back to fix them after they're merged) to do the work to fix up and complete that feature before release (or revert it, which can be messy and would likely lead to far more complaints). That creates a time-constrained deadline, whereas if an unfinished PR *isn't* merged we can simply wait for it to be completed. HR is a great example of this.

Effectively, we can more easily consolidate and organize work on a particular feature if it's kept within the same PR, as opposed to having to keep track of which merged features aren't technically done yet or have bugs. Yes, issues can help with this, but at a certain point you're just duplicating effort.

I think this comes down to the transient nature of a lot of contributions to the game. It's fairly common for someone to PR a feature and then abandon it, which is sometimes not just because it hasn't been merged yet, but because the contributor (as a volunteer) has lost interest and moved on to something else. While this is perfectly fine, as ES is a volunteer project people work on for fun, I think it would create a lot of accountability issues with regards to who is responsible for fixing up a buggy feature originally written by someone who no longer wants to work on it.

**Reply by RisingLeaf** [reply]
Still there is a difference between something being buggy and something not looking quite right yet, of course we don't want to merge stuff that's buggy but maybe we want to merge a boarding rework that doesn't feel quite right yet but is still an improvement.

**Reply by Azure3141** [reply]
I think that still opens the question of who is going to improve it to the expected standards afterward? While it's still a PR there is some obligation on the part of the author to work on it, but after being merged there's not a feasible way to ensure that they finish it before the next release, which could force a situation where a developer has to do the work or the PR has to be messily reverted, spawning complaints.

The only way I can see this working is if the PR author makes a public commitment to prioritize continued work on the feature, (and follow-up PRs) and merging PRs prematurely like that would have to take into account the author's reputation with meeting those commitments and deadlines. That has the potential to produce future drama, but it is a workable solution that might mitigate the other issues.

**Reply by mOctave** [reply]
> I worry that it could lead to situations where we merge buggy / unfinished PRs, and then that creates the obligation for someone ... to do the work to fix up and complete that feature before release

Yes, I can see how this would be a problem, _but_ there's three different reasons why a PR might be merged:

1. It's still buggy. For a code PR, it doesn't work the way it's supposed to or introduces other problems. For a content PR, it is either a grammatical nightmare, broken, or doesn't fit with lore. I totally agree in this situation that it shouldn't be merged.
2. It's not finished. I sort of agree with you that we shouldn't merge unfinished PRs, but at the same time, what counts as finished? I think PRs should be mostly self-contained units that shouldn't be merged until everything actually in the PR is merged. But we shouldn't necessarily avoid merging a PR just because there's other changes that we want, if the PR works as it stands.
3. It's not polished. This is the one that I think _should_ still be merged, at least early in the development cycle. Maybe a code PR creates some weird unintentional behaviour, but only in a documented edge case, and not in a way that changes anything existing. Maybe a UI element has, say, weird horizontal bars across it if you zoom way in. Maybe something could be worded ever so slightly better. Maybe a new image asset is slightly discoloured. But all of those can be fixed later on (if desired), and the person who fixes them doesn't have to be either the original contributor or a dev. It's these polishing changes that I think shouldn't hold up a PR, especially when for most potential contributors, that polishing work is actually really difficult (as can be the case in art and code PRs).

**Reply by Azure3141** [reply]
> Maybe a code PR creates some weird unintentional behaviour, but only in a documented edge case, and not in a way that changes anything existing. Maybe a UI element has, say, weird horizontal bars across it if you zoom way in. Maybe something could be worded ever so slightly better. Maybe a new image asset is slightly discoloured. 
> But all of those can be fixed later on (if desired), and the person who fixes them doesn't have to be either the original contributor or a dev.

I think, as you said, because PRs are mostly self-contained, it opens the question of why they should be merged if there is still polishing needed to make them match the quality of the rest of the game. It isn't necessary for a self-contained PR to be merged to master before fixes can be made to it. Because these small unfinished bits do add up over time, and if it isn't the original contributor, it's a developer or someone else that now has additional work to do in order to polish a PR if they want the game to meet its current standards.

In a case like what you're describing, especially for new people who might not be able to polish their PR to such a high standard, I think the best solution would be for reviewers/developers/contributors to PR that polishing work *to* the contributor's branch before the PR itself is actually merged. That way, helping out with the polishing remains an optional, voluntary burden to take on, (albeit one I think we should definitely encourage more often, especially for newer people) instead of "clean up" work people have to scramble to complete before release.

**Reply by TheGiraffe3** [reply]
This is mainly the view I have as well. If a PR is ready except for just one aspect that would be _nice_ to have, but isn't really necessary, then I think it should just be merged.

Some of the closed PRs are really great ideas, and in most cases the implementation was also good, but ended up being closed because nobody looked at it for three months.

**Comment by xobes** [comment]
Hello, I'm here as a long-time-ago player of Escape Velocity, and I loved it then, as I love endless-sky now.  I recall back then the joy I experienced learning I could tweak the resources using resedit on the mac.  I think plug-ins are essential, and allow for people to do what THEY want and not have to worry about broader consensus ruining their chance to contribute or expand on ideas -- part of the joy is seeing your stuff in the game, so I'm pleased to know plugins are a thing here.

But that's not why I started writing this comment, I got down to where there was some talk about:

> "We have structures and rules here, so they need to be written down. If we did not have them we wouldn't need to write them down, but we have them so it is necessary. This should also include a list of people with statuses above standard contributor."
...
>> "...wiki as a separate [repository](https://github.com/endless-sky/endless-sky-wiki)..."
...
>>> "I don't actually see a problem here, I think the syntax is pretty good, wiki needs some filling up though."

My contribution to this discussion therefore is this:  I have opened a bunch of files so far to look at what's in them.  When it comes to the game-specific files, e.g. .txt files, one can no longer look to the file extension to understand the format and parsing of the file -- it's text, but it's very special text -- and, this is my point -- the file itself doesn't self-document it's syntax (Comments are good documentation in code, for example, and helps a reader understand what's going on in the code) -- could the files (.txt files in particular here) include commentary about how/where they are used/referenced and point to the wiki that describes their individual structure?

Newbies like me would be well served to be able to open a .txt file and see a reference to a wiki explaining how to interpret the file.

Thanks for letting me contribute to the dialog!
